keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,581 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The [kube-storage-version-migrator] supports migrating existing resources to a new storage version but it does not modify a CRD's `status.storedVersions` field after a successful migration, instead requiring that the cluster administrator updates this field manually. As adoption for CRDs continues to increase and new versions of those CRDs are released this design decision will place a non-trivial amount of work on cluster administrators.
+
+Historically, the [kube-storage-version-migrator] has not managed modifying the `status.storedVersion` field because the CRD's storedVersion could change mid-migration unbeknownst to the migrator due to slow/failed watches. This KEP proposes a solution that would allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration, reducing the burden on cluster administrators.
+
+[kube-storage-version-migrator]: https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration.
+- Stop other actors from modifying the storage version while a migration is ongoing.
+
+## Proposal
+
+In order for the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` field after a successful migration, this KEP proposes that the [kube-storage-version-migrator] introduces a [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) that prevents changes to a CRD's `spec.versions[*].storage` fields during a migration using a [fail](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy) failure policy.
+
+The validating admission webhook would need to:
+
+- Intercept all Update operations against CRDs.
+- Prevent changes to a CRD's `spec.versions[*].storage` fields if the CRD includes the `ksvm.sigs.k8s.io/migrating=true` annotation.
+
+A summary of the new migration workflow can be found below:
+
+1. A `StorageVersionMigration` CR for the `foo` CRD has been created, possibly by the trigger-controller as [described here](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/acdee30ced218b79e39c6a701985e8cd8bd33824/USER_GUIDE.md#storage-version-migrator-in-a-nutshell)). 
+2. The [kube-storage-version-migrator] adds the `ksvm.sigs.k8s.io/migrating=true` annotation to the`foo` CRD.
+3. The [kube-storage-version-migrator] begins migrating all `foo` resources as it does today.
+4. When a migration is successful, the `foo` CRD's `status.storedVersion` is patched to only include the new stored version.
```

Comment:
That's a valid concern. We could add a requirement that the admission webhook includes a label, thereby removing the need for a canonical location. Thoughts?

Reply:
Available in the latest version of the PR. Open to changes if requested.

Reply:
Rather than having the ksvm identify if the admission webhook is installed, perhaps a flag that enables / disables this feature is more appropriate, ksvm would simply expect the user to install the webhook made available in the repo, thoughts?